<table>
    <tr>
        <td><img src="https://s3.amazonaws.com/media-p.slid.es/uploads/1485763/images/9060062/Header.png" width="300"/></td>
        <td>&nbsp;</td>
        <td>
            <h1 style="font-size:200%;color:blue;text-align:center">    <FONT COLOR="blue"> Generalidades </p> Parte 1 EDA   </FONT>         </h1></td>         
        <td>
            <tp><p style="font-size:99%;text-align:center">Machine Learning </p></tp>
            <tp><p style="font-size:115%;text-align:center">CoSIAM</p></tp>
            <tp><p style="font-size:115%;text-align:center">Prof. Fabián Sánchez</p></tp>
        </td>
    </tr>
</table>

<FONT SIZE=4 COLOR="Purple"> Objetivo de la sesión 1 </FONT>

En esta sesión se presentarán:

* Introducción al machine learning
* Preprocesamiento de Datos:
    - Carga de datos
    - Exploración de datos (EDA)

**Profesor**: Fabián Sánchez Salazar

**PhD en Matemáticas**

*fabian.sanchez@urosario.edu.co / fabian830348@fkanalytics.com*





# <FONT SIZE=5 COLOR="Purple"> Conceptos Generales </FONT>

A continuación, presentaremos los conceptos de Inteligencia Artificial (AI), Machine Learning (ML) y Deep Learning (DL).


<br>
<center><FONT SIZE=5 COLOR="green"> AI-ML-DL </FONT></center>
<br>

- ***Inteligencia Artificial:*** es una combinación y dispositivos creados con el fin de crear máquinas que imiten capacidades del ser humano.

- ***Machine Learning:*** es un campo de la Inteligencia Artificial que a traves de algoritmos se tratan de identificar patrones en conjuntos de datos masivos y hacer predicciones sobre estos conjuntos.

- ***Deep Learning: *** son algoritmos que usan arquitecturas computacionalmente más complejas , como las redes neuronales, para procesar y hacer predicciones sobre conjuntos de datos.

El siguiente gráfico presenta la relación entre los campos definidos anteriormente.

<center><img src="https://miro.medium.com/v2/resize:fit:720/format:webp/1*H9eMNh36H1gGJXNOxjaeBw.png" alt="centered image" width="400" height="400"></center><center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: https://medium.com/@experiencIA18/diferencias-entre-la-inteligencia-artificial-y-el-machine-learning-f0448c503cd4  </FONT> <figcaption></center>
<br>



## <FONT SIZE=4 COLOR="green"> Machine Learning </FONT>

- El *Machine Learning* o aprendizaje automático de máquina trata de extraer conocimiento de los datos. Es decir, extraer información oculta de los datos.

- Es un campo de investigación que intersecta la estadística, inteligencia artificial y ciencias de la computación.

- Las aplicaciones del *Machine Learning* están presenten en muchos contextos de la vida diaria. Desde recomendaciones automáticas de qué películas para ver, qué comida pedir, que productos comprar hasta el reconocimiento de amigos en tus fotos.

- Al mirar un sitio web complejo como Facebook, Amazon o Netflix, es muy probable que cada parte del sitio contenga múltiples modelos de machine learning.

Dentro del campo del machine learning se estudian algunos tipos de problemas.

## <FONT SIZE=4 COLOR="green"> Aprendizaje supervisado </FONT>

En este caso se desea automatizar el proceso de toma de decisiones a partir de ejemplos conocidos. Estos algoritmos de *Machine Learning* aprenden de los pares input/output conocidos, para crear un **output** para un **input** que nunca ha visto antes sin más ayuda.

Vamos a considerar dos conjuntos de datos:

1. (**Clasificación**) Supongamos que tenemos estos datos

<br>
<center><img src="https://github.com/Fabian830348/cursos/blob/master/supervisado5.png?raw=true" alt="centered image" width="700" height="200"></center>
<br>

donde,

- ***Pregnancies***: Número de embarazos que ha tenido en su vida
- ***Glucose***: Nivel de concentración de glucosa en sangre
- ***BloodPressure***: Presión arterial
- ***SkinThikness***: Espesor de piel a la altura del triceps
- ***Insulin***: Respuesta a dosis de insulina en 2 horas
- ***BMI***: Índice de masa corporal
- ***DiabetesPedigreeFunction***: Presencia de diabetes en ascendencia directa
- ***Age***: Edad del paciente
- ***Outcome***: Variable que queremos predecir:
   - $1$ : Tiene diabetes
   - $0$ : No tiene diabetes

***Problema***: Predecir si una mujer, con ciertos valores en las variables antes menciondas tiene o no diabetes.

2. (**Regresión**) Los siguientes datos contienen la información de en pauta publicitaria en diferentes medios (en miles de dolares)

<center><img src="https://github.com/Fabian830348/cursos/blob/master/supervisado4.png?raw=true" alt="centered image" width="600" height="200"></center>

***Problema***: Predecir las ventas en función de las inversiones que se hacen en pauta publicitaria en ada categoria.

Algunos ejemplos de aplicaciones del aprendizaje supervisado

1. **Clasificación de correos spam**: El modelo de *Machine Learning* utiliza un algoritmo dotado de una gran cantidad de correos electrónicos (los inputs), junto con información sobre si alguno de estos correos electrónicos es spam (objetivo deseado). Dado un nuevo correo electrónico, el algoritmo producirá una predicción sobre si el nuevo correo electrónico es correo no deseado.

2. **Determinar si un tumor es benigno en base a una imagen médica:** La entrada es la imagen y la salida es si el tumor es benigno. Para crear un conjunto de datos para construir un modelo, se necesita una base de datos de imágenes médicas. Asi mismo se necesita la opinión de un experto, por lo que un médico debe observar todas las imágenes y decidir qué tumores son benignos y cuáles no. Incluso podría ser necesario hacer un diagnóstico adicional más allá del contenido de la imagen para determinar si el tumor en la imagen es canceroso o no.

3. **Detección de actividad fraudulenta en transacciones con tarjetas de crédito** La entrada es un registro de la transacción de la tarjeta de crédito y la salida es si es probable que sea fraudulento o no. Suponiendo que usted es la entidad que distribuye las tarjetas de crédito, recopilar un conjunto de datos significa almacenar todas las transacciones y registro si un usuario reporta alguna transacción como fraudulenta.

4. **Predicción del comsumo de tarjeta de crédito. ** Con base a diferentes variables que representan las características de un cliente, se puede predecir su consumo de tarjeta de crédito en un determinado mes o periodo de tiempo.

Por otra parte

## <FONT SIZE=4 COLOR="green"> Aprendizaje No Supervisado </FONT>

Estos son el otro tipo de algoritmo en el que solo se conocen los datos de entrada y no se conoce la salida o etiqueta.

(**Clustering**) En los siguientes datos se muestra información de asesinatos, asaltos, tamaño de la población, violaciones, etc., de estados de EEUU.

<center><img src="https://github.com/Fabian830348/cursos/blob/master/no_supervisado1.png?raw=true" alt="centered image" width="500" height="200"></center>

**Problema**: Identificar grupos de elementos del conjunto que compartan ciertas características.

También se aplica para reducción de dimemsionalidad.

Algunos ejemplos de aplicación:

1. **Segmentación de clientes en grupos con preferencias similares:** Dado un conjunto de registros de clientes, es posible que desee identificar qué clientes son similares, y si hay grupos de clientes con preferencias similares. Para un sitio de compras, estos pueden ser "padres", "ratones de biblioteca" o "jugadores". Porque no conoce de antemano cuáles podrían ser estos grupos, o incluso cuántos hay, no tiene salidas conocidas.

2. **Identificar temas en un conjunto de publicaciones de blog:** Si tiene una gran colección de datos de texto, es posible que desee resumirlos y encontrar temas predominantes en él. Es posible que no sepa de antemano cuáles son estos temas, o cuántos temas puede haber. Por lo tanto, no hay salidas conocidas.

Observe que para entender y poder determinar si un problema es supervisado o no supervisado debemos tener claro el problema que debemos resolver y qué tipos de variables contiene el conjunto de datos.

<center><FONT SIZE=5 COLOR="PURPLE"> MACHINE LEARNING </FONT></center>

<br>
<center><img src="https://github.com/Fabian830348/cursos/blob/master/imagen_ML.png?raw=true" alt="centered image" width="500" height="300"></center><center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: Elaboración propia  </FONT> <figcaption></center>
<br>

Un aspecto importante en el campo del Machine Learning, es que antes de aplicar cualquier modelo, se requiere conocer los datos y alistarlos.

A continuación, revisaremos algunos de estos conceptos.

# <FONT SIZE=5 COLOR="Purple"> Proceso EDA : Exploratory Data Analysis </FONT>

El ***EDA*** es el proceso mediante el cual se exploran los datos e incluye algunos pasos como lo indica el siguiente gráfico.

<center><FONT SIZE=5 COLOR="PURPLE"> PROCESO DE EXPLORACIÓN DE LOS DATOS </FONT></center>

<br>

<center><img src="https://github.com/Fabian830348/cursos/blob/master/imagen_eda.png?raw=true" alt="centered image" width="550" height="350"></center><center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: Elaboración propia </FONT> <figcaption></center>

En el gráfico anterior se muestran algunas etapas fundamentales:

1.	**Preguntas para responder**. Es importante tener claro para qué queremos hacer el procesamiento y análisis de datos. Es decir, qué buscamos responder.

2.	**Generalidades del conjunto de datos**.  Acá buscamos reconocer la dimensión, datos faltantes, y en general un resumen general del dataset.

3.	**Tipos de Variables**. Clasificar las variables en cualitativas y cuantitativas y examinar como se presenta esta información. Por ejemplo, una variable puede estar codificada y verse como numérica, pero ser categórica.

4.	**Estadística Descriptiva**. Se busca resumir la información. Utilizar medidas de tendencia, de dispersión, de forma, de posición, etc., para entender mejor los datos. Por ejemplo, se identifican datos atípicos usando boxplot.

5.	**Visualización**.  Es una etapa fundamental en el análisis de datos. Visualizar la información y poder describirla gráficamente. En esta etapa se usan: diagramas de barras, pie, boxplot, histogramas, líneas de densidad, diagramas de dispersión, entre otros.

6.	**Relación entre variable**. Se explora datos bivariados. Se examinan relaciones entre variables cualitativas (tablas de contingencia) y relaciones entre variables cuantitativas (diagramas de dispersión) se establecen posibles relaciones lineales.

7.	**Reporte de resultados**. Una vez se haga el proceso de exploración y análisis en RStudio se deben reportar los resultados y conclusiones de lo que se encontró en la base de datos.

<center><FONT SIZE=6 COLOR="magenta"> PYTHON </FONT></center>

Python es un lenguaje de programación de propósito general, usado ampliamente para

*  Desarrollo de aplicaciones
*  Desarrollo web
*  Web Scraping
*  **Ciencia de Datos**
*  **Visualización**
*  Aplicaciones a los negocios
*  **Machine Learning**
*  **Deep Learning**

Python cuenta con librerías especializadas dependiendo del tipo de tarea. Algunas de ellas son:

<center><img src="https://github.com/Fabian830348/cursos/blob/master/librerias.png?raw=true" alt="centered image" width="550" height="350"></center> <center><figcaption> <FONT SIZE=1 COLOR="black"> Fuente: https://www.actumlogos.com/python-para-ia.html  </FONT> <figcaption></center>

1.   **Pandas:** manipulación de datos-estadísticas-data.frames
2.   **Matplotlib**: visualización-gráficas
3.   **numpy**: operaciones matemáticas
4.   **spicy**: realizar cálculo simbólico
5.   **scikit-learn**: ejecutar algoritmos de machine learning
6.   **NTKL**: Procesamiento de Lenguaje Natural.
7.   **seaborn**: creación de gráficos estadísticos
8.   **Datetime**: para manipulación de fechas






<center><FONT SIZE=5 COLOR="GREEN"> EXPLORACIÓN DE DATOS EN PYTHON</FONT></center>

<center><img src="https://miro.medium.com/v2/resize:fit:720/format:webp/0*fto5mNElXLZ8TCZr.png" alt="centered image" width="500" height="250"></center>

+ Librería de Python especializada manipulación y análisis de estructuras de datos
+ Permite leer datos en diferentes formatos: csv, excel, bases de datos SQL.
+ Realiza operaciones rápidas y eficientes sobre conjuntos de datos
+ Permite trabajar series de tiempo
+ *import pandas as pd*
+ [Guía Pandas](https://aprendeconalf.es/docencia/python/manual/pandas/)

Los tipos de datos que vamos a trabajar se denominan data.frames

<center><FONT SIZE=5 COLOR="blue"> data.frame </FONT></center>

<center><img src="https://github.com/Fabian830348/cursos/blob/master/dataframe.png?raw=true" alt="centered image" width="600" height="300"></center>

Observe que un data.frame puede tener diferentes tipos de variables

<center><img src="https://github.com/Fabian830348/cursos/blob/master/tiposvariables.png?raw=true" alt="centered image" width="600" height="400"></center>


<FONT SIZE=3 COLOR="magenta"> Algunas funciones para explorar data.frames </FONT>


**df.info()** : Devuelve información (número de filas, número de columnas, índices, tipo de las columnas y memoria usado) sobre el DataFrame df.

**df.shape** : Devuelve una tupla con el número de filas y columnas del DataFrame df.

**df.size** : Devuelve el número de elementos del DataFrame.

**df.columns** : Devuelve una lista con los nombres de las columnas del DataFrame df.

**df.index** : Devuelve una lista con los nombres de las filas del DataFrame df.

**df.dtypes** : Devuelve una serie con los tipos de datos de las columnas del DataFrame df.

**df.head(n)** : Devuelve las n primeras filas del DataFrame df.

**df.tail(n)** : Devuelve las n últimas filas del DataFrame df.

# <FONT SIZE=5 COLOR="Purple"> EJEMPLO DE ANÁLISIS EXPLORATORIO DE LOS DATOS </FONT>

En lo que sigue vamos a trabajar con el conjutno de datos *diabetes*.

Vamos a seguir los siguientes pasos.

## <FONT SIZE=4 COLOR="Green"> Cargar los datos </FONT>

Vamos a cargar los datos desde un repositorio de *github*.

Lo primero que haremos es cargar la librería **pandas**, que nos servirá para manipulación de data.frames.






In [3]:
#para importar la librería pandas.
import pandas as pd

La dirección del conjunto de datos es:

```python
https://raw.githubusercontent.com/Fabian830348/Bases_Datos/master/diabetes.csv
```



In [4]:
url = "https://raw.githubusercontent.com/Fabian830348/Bases_Datos/master/diabetes.csv"

Como los datos están en formato *.csv* usamos la función:

$$pd.read\_csv()$$

In [22]:
diabetes= pd.read_csv(url)

Vamos a trabajar con este último conjunto de datos para hacer una exploración y revisar algunas instrucciones de pandas y gráficas clásicas.

1. Usamos la instrucción **head()** para explorar la cabeza de los datos

- Podemos indicar cuántos datos queremos visualizar simplemente con $head(n)$, sonde $n$ indica el número de registros que queremos ver.

In [ ]:
diabetes.head(7)

2. Usamos la instrucción **tail()** para explorar los últimos datos

In [ ]:
diabetes.tail(6)

3. Con la instrucción **shape** miramos el tamaño de nuestros datos

In [10]:
diabetes.shape

(768, 9)

es decir, el conjunto de datos tiene 768 registros o filas y 9 columnas.

4. Para explorar nombre de las variables o columnas **columns**

In [13]:
diabetes.columns

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')

Recordemos que significan las variables:

- ***Pregnancies***: Número de embarazos que ha tenido en su vida
- ***Glucose***: Nivel de concentración de glucosa en sangre
- ***BloodPressure***: Presión arterial
- ***SkinThikness***: Espesor de piel a la altura del triceps
- ***Insulin***: Respuesta a dosis de insulina en 2 horas
- ***BMI***: Índice de masa corporal
- ***DiabetesPedigreeFunction***: Presencia de diabetes en ascendencia directa
- ***Age***: Edad del paciente
- ***Outcome***: Variable que queremos predecir:
   - $1$ : Tiene diabetes
   - $0$ : No tiene diabetes

5. Podemos ver que tipos de elementos tiene el conjunto de datos con **dtypes**

In [14]:
diabetes.dtypes

Pregnancies                   int64
Glucose                       int64
BloodPressure                 int64
SkinThickness                 int64
Insulin                       int64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object

Es importante verificar que las variables del conjunto de datos hayan sido cargadas en el tipo adecuado, es decir, la variable.

6. **info()** es un resumen de las variables y su clasificación. Además, podemos identificar datos faltantes (NA) en cada una de ellas.

In [ ]:
diabetes.info()

7. Revisaremos si tiene datos faltantes con **isna**

In [ ]:
diabetes.isna()

Esta instrucción indica si en la posición $i,j$ tenemos un valor o un dato faltante. El inconveniente es que nos genera una matriz del mismo tamaño de los datos y si el conjunto es muy grande, no aporta mucha información sobre los NA. Por esa razón, debemos sumar sobre las columnas teniendo en cuenta que:

- False : 0  
- True : 1

In [ ]:
diabetes_na = diabetes.isna().sum()
diabetes_na
#aptos_na[aptos_na>0]

8. Para revisar las estadística generales de las variables cuantitativas usamos **describe()**

In [21]:
diabetes.describe()

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  0.078000   21.000000    0.000000  
25%     27.300000                  0.243750   24.000000    0.000000  
50%     32.000000                  0.372500   29.000000    0.000000  
75%     36.600000                  0.626250   41.000000    1.000000  
max     67.100000                  2.420000   81.000000    1.000000

***Nota:*** Siempre que hagamos una tabla o gráfica es importante indicar qué queremos resaltar sobre la información contenida. Por ejemplo, en la tabla anterior podriamos indicar:

*Se puede observar que el valor mínimo de glucosa reportada es 0 y el máximo es 199. con un promedio de 120*

9. Para acceder a una o varias variable del data.frame podemos usar la instrucción

*datos.["nombre de la variable"]*

veamos algunos ejemplos

In [ ]:
# podemos acceder a una o más variables del data.frame data[]
diabetes["Insulin"]

In [ ]:
# podemos revisar unas estadísticas básicas de una variable numérica con .describe()
diabetes["Insulin"].describe()

También se puede acceder al nombre de una variable asi:
$$datos.variable$$

In [ ]:
diabetes.Insulin

<FONT SIZE=4 COLOR="red"> Tabla de frecuencias </FONT>

Para hacer el conteo de las categorías o clases de una variables podemos usar la instrucción ***.value_counts***

Ejemplo: Para hacer una tabla de frecuencias de la variable *Pregnancies*

In [33]:
freq=diabetes["Pregnancies"].value_counts()
freq

1     135
0     111
2     103
3      75
4      68
5      57
6      50
7      45
8      38
9      28
10     24
11     11
13     10
12      9
14      2
15      1
17      1
Name: Pregnancies, dtype: int64

Si queremos mejorar la presentación, es decir, visualizarlo como un data.frame

In [ ]:
pd.Series.to_frame(freq)

<FONT SIZE=4 COLOR="red"> ALGUNAS GRÁFICAS </FONT>

Para hacer algunas gráficas de las variables usaremos las siguientes librerías:

In [29]:
import matplotlib.pyplot as plt
import plotly.express as px
import seaborn as sb

In [ ]:
# gráfico de barras
px.bar(freq,
       color = freq.index)

In [ ]:
# diagrama de tortas
px.pie(freq,
       values= freq,
       names = freq)

In [38]:
# histograma
px.histogram(diabetes.BloodPressure)

In [39]:
fig = px.histogram(diabetes.BloodPressure,                            # variable
                   nbins = 20,                              # número de intervalos
                   x = "BloodPressure",                            # nombre de la variable x
                   title = "Histograma con edición"         # título del gráfico
                                      )

fig.update_traces(marker_color = 'cyan',                    # color de relleno
                  marker_line_color = 'red',                # color del borde
                  marker_line_width = 1,                    # grueso de la línea
                  opacity = 1)                              # transparencia del color

fig.show()

In [40]:
# Diagrama de cajas
px.box(diabetes.Insulin)

In [41]:
px.box(diabetes.Insulin,
       color = diabetes.Pregnancies)

<FONT SIZE=4 COLOR="red"> HERRAMIENTA PARA VISUALIZACIÓN </FONT>



In [ ]:
!pip install sweetviz

In [48]:
import sweetviz as sw

In [49]:
reporte=sw.analyze(diabetes, target_feat= "Outcome")

                                             |          | [  0%]   00:00 -> (? left)

In [ ]:
from IPython.display import HTML
reporte.show_html()
HTML("SWEETVIZ_REPORT.html")